### PR TITLE
LetPoly1: Fix parametric polymorphism in the type checker

### PR DESCRIPTION
**Pull request by konnov**
This PR fixes the type checker, in order to support polymorphism as explained in the issue #869. It does not fix further preprocessing steps, which still produce errors on polymorphic definitions. Since this PR contains a lot of changes already, I have decided to stop here.

The main changes are done in `TypeUnifier` and `EtcTypeChecker`. Regarding the type unifier, I basically had to rewrite it, in order to work correctly on substitutions and terms that may contain free variables. To this end, the unifier works with equivalence classes of variables, instead of individual variables. The main focus was on soundness, rather than efficiency, so the new code may be less efficient than the old one.

In the type checker, the main changes are made in the processing of `EtcAppByName` and `EtcLet`, as expected. The behavior should be now consistent with the memo on the type checker.

I know that this PR is relatively large. It's because the changes in the type checker pulled the necessary updates to the type unifier, substitutions, etc.

- [x] Tests added for any new code
- [x] Ran `make fmt-fix` (or had formatting run automatically on all files edited)
- [ ] Documentation added for any new functionality
- [x] Entry added to [UNRELEASED.md](./UNRELEASED.md) for any new functionality

### CODE REVIEW

**Comment by konnov** [comment]
It looks like we are running into non-determinism in the test for `ConfigUnsorted.tla` :(

**Comment by konnov** [comment]
I have fixed an elusive bug, where the type checker was overriding the inferred signature with the signature in a type annotation. This made the operator body and the operator type inconsistent in the quantified variables.

**Review by Kukovec — APPROVED** [review]
I have some minor comments on documentation, and a more general one that didn't fit any particular line. It seems to me that your implementation of equivalence classes is a bit ad-hoc. Specifically, you reference the invariant that the classes are disjoint (which they should be), but the class design itself does not guarantee this. Perhaps you should implement a disjoint-set data structure, canonically used to represent equivalence classes (https://en.wikipedia.org/wiki/Disjoint-set_data_structure). It should take care of disjointness automatically, and it supports the two operations you care about: lookup and class merging.

**Comment by konnov** [comment]
> I have some minor comments on documentation, and a more general one that didn't fit any particular line. It seems to me that your implementation of equivalence classes is a bit ad-hoc. Specifically, you reference the invariant that the classes are disjoint (which they should be), but the class design itself does not guarantee this. Perhaps you should implement a disjoint-set data structure, canonically used to represent equivalence classes (https://en.wikipedia.org/wiki/Disjoint-set_data_structure). It should take care of disjointness automatically, and it supports the two operations you care about: lookup and class merging.

Yes, as we have discussed, it is an ad hoc implementation of equivalence classes. Let's plan refactoring with union-find after this PR has been merged. Since you have a clear idea of how to do that, it would be great if you could take the refactoring token.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/EqClass.scala [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+   */
+  def copy(): EqClass = {
+    EqClass(this.typeVars)
+  }
+
+  def canEqual(other: Any): Boolean = other.isInstanceOf[EqClass]
+
+  override def equals(other: Any): Boolean = other match {
+    case that: EqClass =>
+      (that canEqual this) &&
```
[the remark this answers is not in the document]
Comment (konnov) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/EqClass.scala:36]: True. I think it was generated by Intellij :)

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/TypeContext.scala [hunk 2]
```diff
@@ -5,20 +5,22 @@ import at.forsyte.apalache.tla.lir.TlaType1
 import scala.collection.immutable.SortedMap
 
 /**
- * A type context assigns poly- or monotypes to names (such as operator names and the names of bound variables).
+ * A type context assigns poly- or monotypes to names (such as operator names and the names of quantified variables).
  *
- * @author
+ * @param poolSize the largest variable index that should be treated as non-free, that is, not universally quantified
```
[the remark this answers is not in the document]
Comment (konnov) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/TypeContext.scala:10]: True. The name `poolSize` is counterintuitive. I have renamed it to `largestVarIndex`.

**Comment by konnov** [comment]
I will merge this branch in unstable. So we can integrate disjoint sets in a later PR.

**Pull request merged**